runtime/nvqir/custatevec/CuStateVecCircuitSimulator.cu

```diff
@@ -77,6 +83,252 @@ __global__ void setFirstNElements(T *sv, const T *__restrict__ sv2, int64_t N) {
+      cudaSetDevice(currentDevice);
+  }
+
+  void extractValues(std::complex<ScalarType> *value, std::size_t start,
+                     std::size_t end) const {
+    // FIXME Wait for Brunos PR on this, should be able to just
+    // get the element with cudaMemcpy
+    // cudaMemcpy(&value,
+    //            reinterpret_cast<std::complex<ScalarType> *>(devicePtr) + idx,
+    //            sizeof(std::complex<ScalarType>), cudaMemcpyDeviceToHost);
+    cudaDataType_t cuStateVecCudaDataType = CUDA_C_64F;
+    if constexpr (std::is_same_v<ScalarType, float>) {
+      cuStateVecCudaDataType = CUDA_C_32F;
+    }
+
+    checkAndSetDevice();
+
+    custatevecAccessorDescriptor_t accessor;
+    const uint32_t nIndexBits = std::log2(size);
+    // Note: we use MSB bit ordering when reporting the state vector
+    // hence, bit ordering vector = [N-1, N-2, ..., 0]
+    std::vector<int32_t> bitOrdering(nIndexBits);
+    std::iota(std::rbegin(bitOrdering), std::rend(bitOrdering), 0);
+    std::size_t extraWorkspaceSizeInBytes = 0;
+    // create accessor view
+    HANDLE_ERROR(custatevecAccessorCreateView(
+        handle, devicePtr, cuStateVecCudaDataType, nIndexBits, &accessor,
+        bitOrdering.data(), bitOrdering.size(),
+        /*maskBitString*/ nullptr, /*maskOrdering*/ nullptr,
+        /*maskLen*/ 0, &extraWorkspaceSizeInBytes));
+    // allocate external workspace if necessary
+    void *extraWorkspace = nullptr;
+    if (extraWorkspaceSizeInBytes > 0)
+      HANDLE_CUDA_ERROR(cudaMalloc(&extraWorkspace, extraWorkspaceSizeInBytes));
+
+    // set external workspace
+    HANDLE_ERROR(custatevecAccessorSetExtraWorkspace(
+        handle, accessor, extraWorkspace, extraWorkspaceSizeInBytes));
+
+    HANDLE_ERROR(
+        custatevecAccessorGet(handle, accessor, &value[0], start, end));
+    // destroy descriptor
+    HANDLE_ERROR(custatevecAccessorDestroy(accessor));
+    // free extra workspace if allocated
+    if (extraWorkspaceSizeInBytes > 0)
+      HANDLE_CUDA_ERROR(cudaFree(extraWorkspace));
+  }
+
+  bool isDevicePointer(void *ptr) const {
+    cudaPointerAttributes attributes;
+    cudaPointerGetAttributes(&attributes, ptr);
+    return attributes.type > 1;
+  }
+
+  std::size_t deviceFromPointer(void *ptr) const {
+    cudaPointerAttributes attributes;
+    cudaPointerGetAttributes(&attributes, ptr);
+    return attributes.device;
+  }
+
+public:
+  CusvState(std::size_t s, void *ptr) : size(s), devicePtr(ptr) {
+    custatevecCreate(&handle);
+  }
+
+  CusvState(const std::vector<std::size_t> &shape,
+            const std::vector<std::complex<double>> &data) {
+    if (shape.size() != 1)
+      throw std::runtime_error(
+          "CusvState must be created from data with 1D shape.");
+
+    size = shape[0];
+    cudaMalloc((void **)&devicePtr, size * sizeof(std::complex<ScalarType>));
+    cudaMemcpy(devicePtr, &data[0],
+               data.size() * sizeof(std::complex<ScalarType>),
+               cudaMemcpyHostToDevice);
```

Comment:
```suggestion
    HANDLE_CUDA_ERROR(cudaMemcpy(devicePtr, &data[0],
               data.size() * sizeof(std::complex<ScalarType>),
               cudaMemcpyHostToDevice));
```

Reply:
Also, just double check that it was intentional:  `data` (`std::vector<std::complex<double>>`) is being copied as a `std::complex<ScalarType>*`. Could `ScalarType` be `float`?